src/parser/generated_expression_visitor.cpp

```diff
@@ -0,0 +1,77 @@
+#include "duckdb/parser/generated_expression_visitor.hpp"
+
+namespace duckdb {
+
+//===--------------------------------------------------------------------===//
+// Base visitor
+//===--------------------------------------------------------------------===//
+
+void GeneratedExpressionVisitor::ExtractLambdaParameters(ParsedExpression &expr) {
+	if (expr.type == ExpressionType::COLUMN_REF) {
+		auto &column_ref = (ColumnRefExpression &)expr;
+		D_ASSERT(!column_ref.IsQualified());
+		auto &col_names = column_ref.column_names;
+		D_ASSERT(col_names.size() == 1);
+		lambda_parameters.insert(col_names[0]);
+	}
+	ParsedExpressionIterator::EnumerateChildren(
+	    expr, [&](const ParsedExpression &child) { this->ExtractLambdaParameters((ParsedExpression &)child); });
+}
+
+void GeneratedExpressionVisitor::VisitExpression(ParsedExpression &expr) {
```

Comment:
For generated columns we need to apply certain modifications to the saved expression, they all share a common route, which I have moved to VisitExpression.
VisitExpression calls Visit which is virtual, and encapsulates the specific modification that needs to be made.

As part of this common route, they all need to be aware of lambda parameters, so it does not treat those column references as regular column references, so GeneratedExpressionVisitor also defines the ExtractLambdaParameters function which it uses on the left hand side of the lambda expression, to figure out which column references are part of a lambda.

That being said: it's quite naive, if there's a lambda reference "x" it will treat every "x" column reference as that lambda parameter and skip it

Also, it does do duplicate work, because they all separately walk the expression and perform all the VisitExpression steps separately.
I think they get called in different parts of the code, I'm not sure if they could be combined.
Or maybe I could cache some of this for the same expression